### ISSUE TITLE
[question] Extending settings and options with new `python_requires` syntax

### ISSUE THREAD

**Issue by DoDoENT** [conan-io/conan#6612]
I am trying to migrate my recipes to use the new `python_requires` syntax as described [in the documentation](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#python-requires).

However, I cannot find a way to correctly extend or override options and settings from the base conanfile.

In the old `python_requires`, I used to do this:

```python
from conans import python_requires, tools

base = python_requires('MyBaseConanFile/1.0.0@company/stable')

class MyPackage(base.MyBaseConanFile):
    settings = ("os_build", "arch_build") + base.MyBaseConanFile.settings
    options = dict(base.MyBaseConanFile.options, **{
        'another_option': [True, False]
    })
    default_options = dict(base.MyBaseConanFile.default_options, **{
        'another_option': True
    })
```

How to achieve the same with the new syntax?

The following does not work:

```python

class MyPackage(ConanFile):
    python_requires = 'MyBaseConanFile/1.0.0@company/stable'
    python_requires_extend = "MyBaseConanFile.MyBaseConanFile"
 
    options = dict(self.python_requires_extend['MyBaseConanFile.MyBaseConanFile'].options, **{
        'another_option': [True, False]
    })
```

The `self` is not available in class attribute context. The documentation says nothing about that case. Is this even possible with the new syntax?

**Comment by memsharded** [comment]
No, this is not possible.
While it is possible to inherit them, modifying them in that way, it is not possible, as the typical inheritance mechanisms only allow overwrite the base class attributes, not operating on them.

The main purpose of the new ``python_requires`` was to recover from the inversion of control, because it is quite complex to maintain.

I have submitted a exploratory possible solution in https://github.com/conan-io/conan/pull/6614, as a starting point to discuss.

**Comment by DoDoENT** [comment]
Maybe we could have some class-level attributes indicating the policies what should be done with inherited attributes.

For example:

```python
class MyPackage(ConanFile):
    python_requires = 'MyBaseConanFile/1.0.0@company/stable'
    python_requires_extend = "MyBaseConanFile.MyBaseConanFile"
    python_requires_options_policy = 'merge' # or 'override', which is the current default
    python_requires_default_options_policy = 'override'
    python_requires_settings_policy = 'merge'
 
    # the options can now be a clean dictionary, without the dict( ..., ** ) hack
    options = {
        'another_option': [True, False]
    }

   # in override mode, so we need to set default options for both inherited and current options
   default_options = {
       'base_option': True,
       'another_option': False
```

The issue here is that we would need a policy for each specific attribute, which is horrible. 

Your proposal in #6614 is even more flexible, as in `python_require_extend_class` it would be possible to perform arbitrary operations on all class attributes (including those for which documentation states that are not currently inherited, such as `short_paths`) - not just merging and overriding. I like the idea, but I have an eery feeling that this will be a cause of lots of exciting new bugs.

**Comment by DoDoENT** [comment]
Just a side note: Until this gets resolved, I managed to create a forward-compatible base conanfile which works with both the current version of `python_requires` and should work with the new syntax:

Base conanfile:
```python
class MyBaseConanFile(object):
    options = {
        'base_option': [True, False]
    }
    settings = "os", "compiler", "build_type", "arch"

class MyBaseConanFilePkg(ConanFile):
    name = "MyBaseConanFile"
    version = "1.0.0"
```

and the usage in current models is as follows (notice the multiple inheritance):

```python

from conans import python_requires, ConanFile

base = python_requires('MyBaseConanFile/1.0.0@company/stable')

class MyPackage(base.MyBaseConanFile, ConanFile):
    settings = ("os_build", "arch_build") + base.MyBaseConanFile.settings
    options = dict(base.MyBaseConanFile.options, **{
        'another_option': [True, False]
    })
    default_options = dict(base.MyBaseConanFile.default_options, **{
        'another_option': True
    })
```

and the (expected) future usage would be something like:

```python
class MyPackage(ConanFile):
    python_requires = 'MyBaseConanFile/1.0.0@company/stable'
    python_requires_extend = "MyBaseConanFile.MyBaseConanFile"
 
    @classmethod
    def python_require_extend_class(cls, c):
        cls.options = c.options + cls.options

    options = {
        'another_option': [True, False]
    }
```

**Comment by jgsogo** [comment]
I've posted a different workaround here (https://github.com/conan-io/conan/pull/6614#issuecomment-595088707), it is a bit misleading because of the dynamic inheritance behind the scenes that Conan is doing to inject the python_requires_extend as a base class, but it works :man_shrugging:

**Comment by memsharded** [comment]
``init()`` method for this will be released in Conan 1.24
